server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -630,6 +699,18 @@ private void balanceByWeights(NodeSorter sorter) {
                             sorter.sort(0, relevantNodes);
                             lowIdx = 0;
                             highIdx = relevantNodes - 1;
+
+                            if (routingNodes.getRelocatingShardCount() > 0) {
+                                // ES-12955: Check routingNodes.getRelocatingShardCount() > 0 in case the first relocation is a THROTTLE.
+                                // This should rarely happen since in most cases, we don't throttle unless there is an existing relocation.
+                                // But it can happen in production for frozen indices when the cache is still being prepared. It can also
+                                // happen in tests because we have decider like RandomAllocationDecider that can randomly return THROTTLE
+                                // when there is no existing relocation.
+                                shardBalanced = true;
+                            }
+                            if (completeEarlyOnShardAssignmentChange && shardBalanced) {
+                                return true;
+                            }
```

Comment:
This is the fix. See also 70b99deb0fd49cf8ff8c8635fe7c1928e2fd59fe for its commit.

There were two issues:
1. The main one is the we should only flag `shardBalanced` to `true` when there is shard movement on the cluster instead of just on the model node.
2. The second issue is that in production, we can actually balance being throttled without any prior shard movement due to [HasFrozenCacheAllocationDecider](https://github.com/elastic/elasticsearch/blob/3f56dbc80ae4625895e659de5b420b12f751cf89/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/decider/HasFrozenCacheAllocationDecider.java#L86). I updated the comment to reflect it.